Lesson 5 (#5)
ReDI-School/js-berlin-2020-fall: lesson5.md

```js
if (hasReservation && !isWearingShorts) {
  console.log("can enter");
} else {
  console.log("cannot enter");
}
```

---

### Quiz

A developer must know C++ or must know HTML, CSS and JavaScript to apply.

Can you write the `if` statement to figure out whether someone can apply based
on the variables `knowsCpp`, `knowsHTML`, `knowsCSS` and `knowsJavaScript`?

```js
if (knowsCpp || (knowsHTML && knowsCSS && knowsJavaScript)) {
  console.log("can apply");
}
```
<!-- .element: class="fragment" -->

---

Harald knows C++, HTML and JavaScript. Can he apply?

```js
if (knowsCpp || (knowsHTML && knowsCSS && knowsJavaScript)) {
  console.log("can apply");
}
```

```js
// step 1 - replace variables with values:
if (true || (true && false && true))
 // step 2 - compute the && in the parentheses
if (true || false)
 // step 3 - compute the ||
if (true)
```
<!-- .element: class="fragment" -->

---

### Exercise

We are implementing a simple program that calculates the bonus an employee will get at the end of the year.

The company will give 1€ bonus for each regular work day, and 3€ for each extra work day.

---

(regular day: 1€, extra: 3€)

> The employee gets, depending on the bonus:
> * a pay raise, if more than 250€ bonus.
> * nothing, if less than 100€.
> * a chocolate box, for any other bonus.

Reviewer: ```suggestion
The employee gets:
* a pay raise, if the total bonus is more than 250€.
* nothing, if the total bonus is less than 100€.
* a chocolate box, for any other bonus.
```

```js
function multiplyBy4(x) {
  return x * 4;
}

let number = multiplyBy4(7);
console.log(number); // 28

number = multiplyBy4(number);
console.log(number);
// what is the output here?
```

Answer: 112
<!-- .element: class="fragment" -->

---

multiple parameters are also possible:

```js
function multiply(x, y) {
  return x * y;
}

let number = multiply(2, 3);
console.log(number); // 6

number = multiply(10, number);
console.log(number);
// what is the output here?
```

Answer: 60
<!-- .element: class="fragment" -->

---

### How to create a function

We write:
* `function`
* name of the function
* `(param1, param2, ...)`
  * parameters are just like variables!
* `{ ... }` the block where we have our code
  * we might add a `return` at the end, depending on what we need.

---

```js
function nameOfFunction (param1, param2, ...) {
  // any code you want!
  // even if-else statements!
  // anything!

  return value; // return is optional, depends on what you need

  // what happens if we don't have a return??
}
```

Functions without a return statements return undefined
<!-- .element: class="fragment" -->

---

### calling a function

> to execute a function, we write the function name, and then `()`, we pass any parameters we want into the `()`
> 
> ```js
> function sayHello(){
>   console.log("Hello!");
> }
> 
> // the function sayHello
> // does not require any parameters
> // that is why the parenthesis () are empty.

Reviewer: ```suggestion
to execute a function, we write the function name, and then `()`, we pass any values we want into the `()`

```js
function sayHello(){
  console.log("Hello!");
}

// the function sayHello
// does not require any values, it has no parameters
// that is why the parenthesis () are empty.
```

sayHello();
```

---

### calling a function

```js
function greet(name){

>   console.log("Hello " + name + "!");

Reviewer 2: Have we introduced string concatenation yet? (I don't know)

Author: https://redi-school.github.io/js-berlin-2020-fall/#/29 :)

}

// the function greet takes one parameter: name
// that is why we write "Bob" in the parenthesis.
// we are giving the value "Bob" to the function.
greet("Bob");
// "Hello Bob!"
```

---

### calling a function

```js
function square(x) {
  return x * x;
}

// the function 'square' takes one parameter: x
// we have to give it a value in the parenthesis ()
// in this example, the value is 5
// however, the function 'square' also returns a value
// we save the returned value in a variable to use it later
let numberSquared = square(5);

console.log(numberSquared); // 25
```

---

### Exercise

Write a function called `celsiusToFahrenheit`, which takes the degrees in C, and returns it in F.

test your function with the values 28 and 31.

Remember:

Fahrenheit = Celsius multiplied by 1.8 plus 32

---

### Solution

```js
function celsiusToFahrenheit(degreesCelsius) {
  let degreesFahrenheit = degreesCelsius * 1.8 + 32;
  return degreesFahrenheit;
}

console.log("28 degrees: " + fahrenheitToCelsius(28));
console.log("32 degrees: " + fahrenheitToCelsius(32));
```

---

### Exercise

> Write a `min` function that returns the smaller of the two parameters:

Reviewer: ```suggestion
Write a `min` function that returns the smaller of the two numbers:
```

```js
min(1, 2);    // should return 1
min(100, 99); // should return 99
min(-10, 0);  // should return -10
```

---

### Solution

```js
function min(number1, number2) {
  if (number1 <= number2) {
    return number1;
  } else {
    return number2;
  }
}

> console.log('min of 1 and 2 is ' + min(1, 2));

Reviewer 2: Before you used " for your strings

Author: Just to remind people that both " and ' are valid :) https://redi-school.github.io/js-berlin-2020-fall/#/23

Reviewer 2: ok.

Other point: are we explaining why we can `+` a string with a number? I think we haven't done that before. Do we say something about type coercion in general at this point?

Author: We told the students not to do it ;) OK - I'll mention it
